src/test/java/org/broadinstitute/hellbender/utils/read/ArtificialBAMBuilderUnitTest.java

```diff
@@ -0,0 +1,96 @@
+package org.broadinstitute.hellbender.utils.read;
+
+import htsjdk.samtools.SAMRecord;
+import htsjdk.samtools.SamReader;
+import htsjdk.samtools.SamReaderFactory;
+import org.broadinstitute.hellbender.utils.test.BaseTest;
+import org.testng.Assert;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+
+public final class ArtificialBAMBuilderUnitTest extends BaseTest {
+    @DataProvider(name = "ArtificialBAMBuilderUnitTestProvider")
+    public Object[][] makeArtificialBAMBuilderUnitTestProvider() {
+        final List<Object[]> tests = new LinkedList<>();
+
+        final List<Integer> starts = Arrays.asList(
+                1, // very start of the chromosome
+                ArtificialBAMBuilder.BAM_SHARD_SIZE - 100, // right before the shard boundary
+                ArtificialBAMBuilder.BAM_SHARD_SIZE + 100 // right after the shard boundary
+        );
+
+        for ( final int readLength : Arrays.asList(10, 20) ) {
+            for ( final int skips : Arrays.asList(0, 1, 10) ) {
+                for ( final int start : starts ) {
+                    for ( final int nSamples : Arrays.asList(1, 2) ) {
+                        for ( final int nReadsPerLocus : Arrays.asList(1, 10) ) {
+                            for ( final int nLoci : Arrays.asList(10, 100, 1000) ) {
```

Comment:
More heavily nested combinatorial testing -- can you confirm that the runtime is not too long?